File: clients/python/telemetry_client.py
import socket
import threading
from typing import Callable, Dict, Optional
from status_parser import parse

Listener = Dict[str, Callable]

class TelemetryClient:
    """
    Cliente TCP simple con un hilo receptor.
    Callbacks esperadas en 'listener':
      - on_connected()
      - on_login_result(ok: bool, line: str)
      - on_status(kv: Dict[str, str])
      - on_server_message(line: str)
      - on_error(msg: str, ex: Exception | None)
      - on_closed()
    """

    def __init__(self, host: str, port: int, role: str, user: str, password: str, listener: Listener):
        self.host = host
        self.port = port
        self.role = role
        self.user = user
        self.password = password
        self.listener = listener

        self._sock: Optional[socket.socket] = None
        self._send_lock = threading.Lock()
        self._recv_thread: Optional[threading.Thread] = None
        self._stop_flag = threading.Event()

    def is_connected(self) -> bool:
        return self._sock is not None

    def connect_async(self) -> None:
        t = threading.Thread(target=self._run, name="net-recv", daemon=True)
        t.start()
        self._recv_thread = t

    def _run(self) -> None:
        try:
            self._sock = socket.create_connection((self.host, self.port), timeout=15.0)
            self._sock_file = self._sock.makefile("r", encoding="utf-8", newline="\n")
            self._call('on_connected')

            # LOGIN
            login_line = f"LOGIN {self.role} {self.user} {self.password}\n"
            self.send(login_line, raw=True)

            # Primera línea
            first = self._sock_file.readline()
            ok = bool(first) and first.upper().startswith("OK")
            is_telemetry = bool(first) and (first.startswith("STATUS ") or first.startswith("DATA "))
            if not is_telemetry:
                self._call('on_login_result', ok, first.strip() if first else "(sin respuesta)")
            else:
                self._call('on_status', parse(first))

            # Loop receptor
            while not self._stop_flag.is_set():
                line = self._sock_file.readline()
                if not line:
                    break
                line = line.rstrip("\r\n")
                if line.startswith("STATUS ") or line.startswith("DATA "):
                    self._call('on_status', parse(line))
                else:
                    self._call('on_server_message', line)

        except Exception as ex:
            self._call('on_error', "Fallo de conexión/IO", ex)
        finally:
            self.close()

    def send(self, text: str, raw: bool = False) -> None:
        if not self._sock:
            return
        data = text if raw else (text + "\n")
        with self._send_lock:
            try:
                self._sock.sendall(data.encode("utf-8"))
            except Exception as ex:
                self._call('on_error', "Fallo enviando datos", ex)

    def close(self) -> None:
        self._stop_flag.set()
        try:
            if hasattr(self, "_sock_file") and self._sock_file:
                self._sock_file.close()
        except Exception:
            pass
        try:
            if self._sock:
                self._sock.close()
        except Exception:
            pass
        finally:
            self._sock = None
            self._call('on_closed')

    # Helper para llamar callbacks si existen
    def _call(self, name: str, *args):
        fn = self.listener.get(name)
        if callable(fn):
            try:
                fn(*args)
            except Exception:
                pass

File: clients/java/src/main/java/com/telematica/client/TelemetryClient.java
package com.telematica.client;

import javax.swing.*;
import java.io.*;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.Map;

public class TelemetryClient {
    public interface Listener {
        void onConnected();
        void onLoginResult(boolean ok, String line);
        void onStatus(Map<String, String> kv);
        void onServerMessage(String line);
        void onError(String msg, Exception ex);
        void onClosed();
    }

    private final String host; private final int port; private final String role; private final String user; private final String pass;
    private final Listener listener;

    private volatile Socket socket; private volatile BufferedReader in; private volatile PrintWriter out;

    public TelemetryClient(String host, int port, String role, String user, String pass, Listener listener) {
        this.host = host; this.port = port; this.role = role; this.user = user; this.pass = pass; this.listener = listener;
    }

    public String getRole() {
        return role;
    }
    public boolean isObserver() {
        return role != null && role.equalsIgnoreCase("OBSERVER");
    }
    public boolean isAdmin() {
        return role != null && role.equalsIgnoreCase("ADMIN");
    }

    public boolean isConnected() { return socket != null && socket.isConnected() && !socket.isClosed(); }

    public void connectAsync() {
        new Thread(() -> {
            try {
                socket = new Socket(host, port);
                in = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8));
                out = new PrintWriter(new OutputStreamWriter(socket.getOutputStream(), StandardCharsets.UTF_8), true);

                SwingUtilities.invokeLater(listener::onConnected);

                // LOGIN
                String loginLine = String.format("LOGIN %s %s %s", role, user, pass);
                out.println(loginLine);

                // Primera línea: puede ser OK/ERR o telemetría (STATUS/DATA)
                String first = in.readLine();
                boolean ok = first != null && first.toUpperCase().startsWith("OK");
                boolean isTelemetry = first != null && (first.startsWith("STATUS ") || first.startsWith("DATA "));
                if (!isTelemetry) {
                    final String f = first != null ? first : "(sin respuesta)";
                    SwingUtilities.invokeLater(() -> listener.onLoginResult(ok, f));
                } else {
                    Map<String, String> kv = StatusParser.parse(first);
                    SwingUtilities.invokeLater(() -> listener.onStatus(kv));
                }

                // Loop receptor
                String line;
                while ((line = in.readLine()) != null) {
                    final String l = line;
                    if (l.startsWith("STATUS ") || l.startsWith("DATA ")) {
                        Map<String, String> kv = StatusParser.parse(l);
                        SwingUtilities.invokeLater(() -> listener.onStatus(kv));
                    } else {
                        SwingUtilities.invokeLater(() -> listener.onServerMessage(l));
                    }
                }
            } catch (Exception ex) {
                SwingUtilities.invokeLater(() -> listener.onError("Fallo de conexión/IO", ex));
            } finally {
                closeQuietly();
                SwingUtilities.invokeLater(listener::onClosed);
            }
        }, "net-recv").start();
    }

    public void send(String text) {
        if (out != null) out.print(text);
        if (out != null) out.flush();
    }

    private void closeQuietly() {
        try { if (in != null) in.close(); } catch (IOException ignored) {}
        try { if (out != null) out.close(); } catch (Exception ignored) {}
        try { if (socket != null) socket.close(); } catch (IOException ignored) {}
    }
}
